google-cloud-tools-plugin/src/com/google/gct/idea/debugger/ServerToIDEFileResolver.java

```diff
@@ -0,0 +1,138 @@
+
+  /**
+   * Utility method that returns the full class name for a file.
+   */
+  public static String getCloudPathFromJavaFile(PsiJavaFile javaFile) {
+    return javaFile.getPackageName().replace('.', '/') + "/" + javaFile.getName();
+  }
+
+  /**
+   * Given a possible file path, returns a VirtualFile instance that the IDE can work with.
+   *
+   * <p>This method tries to fetch a file in three ways. First, it uses its full path in the local
+   * file system. Then, it tries the full class name form (com/google/gct/idea/debugger/
+   * CloudDebugProcess.java). Finally, it searches for possible file matches within the project.
+   */
+  public VirtualFile getFileFromPath(@NotNull Project project, @NotNull String path) {
+    // Try the relative full project path.
+    VirtualFile file = FILE_SYSTEM.findFileByPath(project.getBasePath() + "/" + path);
+    // Try class name with package and class file name.
+    if (file == null) {
+      PsiClass[] matchingClasses = JavaPsiFacade.getInstance(project)
+          .findPackage(getPackageFromPath(path))
+          .findClassByShortName(getClassNameFromPath(path), GlobalSearchScope.allScope(project));
+      if (matchingClasses.length > 0) {
+        file = matchingClasses[0].getContainingFile().getVirtualFile();
+      }
+    }
+    // If we still couldn't find the file, search for possible file name matches and return the
+    // first.
+    // We might want to improve string matching and return more than one possible match.
+    if (file == null) {
+      Collection<VirtualFile> projectJavaFiles = FileBasedIndex.getInstance().getContainingFiles(
+          FileTypeIndex.NAME, JavaFileType.INSTANCE, GlobalSearchScope.projectScope(project));
+      for (VirtualFile projectFile : projectJavaFiles) {
+        if (projectFile.getName().equals(path)) {
+          file = projectFile;
+          break;
+        }
+      }
+    }
+    return file;
+  }
+
+  /**
+   * Produces a Java package name from a file path.
+   *
+   * <p>Example: returns "com.java.package" from "com/java/package/Class.java".
+   */
+  @VisibleForTesting
+  static String getPackageFromPath(String path) {
+    String[] tokens = path.split("/");
+    StringBuilder packageBuilder = new StringBuilder();
+    if (tokens.length > 1) {
+      if(tokens[0].length() > 0) {
+        packageBuilder.append(tokens[0]);
+      }
+      for (int nToken = 1; nToken < tokens.length - 1; nToken++) {
+        if(tokens[nToken].length() > 0 && packageBuilder.length() > 0) {
+          packageBuilder.append(".");
+        }
+        packageBuilder.append(tokens[nToken]);
+      }
+    }
+
+    return packageBuilder.toString();
+  }
+
+  /**
+   * Produces a class name from a file path.
+   *
+   * <p>Example: returns "Class" from "com/java/package/Class.java".
+   */
+  @VisibleForTesting
+  static String getClassNameFromPath(String path) {
+    if (path.indexOf('.') != -1 && path.lastIndexOf("/") < path.lastIndexOf(".")) {
```

Comment:
Because the path can't be a class? I'm trying to be as flexible as possible, to accept anything that CDB sends. For breakpoints, CDB tries to match whatever path the client sends and returns that same path. It might be possible that a path is defined as /foo/bar/Class and CDB matches it to a file, in which case this support would be necessary.